[CUMULUS-2170] Add FilterExpression to granule scan

dopeters opened the pull request:
**Summary:** Makes Granule Inventory Reports filterable

Addresses [CUMULUS-2170: Update reconciliationReport with Granule Inventory type endpoint to take filter parameters](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2170)

## Changes

* Pass searchParams to `granuleAttributeScan` when creating a granule inventory report
* Use searchParams to create FilterExpression for DynamoDbSearchQueue
* Made a separate function to handle creating the params to pass to the search queue so that it could be reused.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [ ] Integration tests

kkelly51 commented on packages/api/lambdas/reports/granule-inventory-report.js:
```diff
@@ -38,6 +30,24 @@ async function createGranuleInventoryReport(recReportParams) {
+  const searchParams = {};
+  fields.forEach((field) => {
+    let searchField = field;
+    if (field === 'granuleUr') searchField = 'granuleId';
```
Are they not named the same thing?

dopeters replied:
The csv column is called `granuleUr`, while the API accepts `granuleId`. Didn't want to change the output, but wanted to use the fields as an easy way of getting relevant search params, so just had it replace the name here.

jennyhliu commented on packages/api/lib/reconciliationReport.js:
```diff
@@ -117,6 +117,30 @@ function convertToDBGranuleSearchParams(params) {
+function convertToDBScanGranuleSearchParams(params) {
+  const {
+    collectionId,
+    granuleId,
```
Is the `granuleId` from the `search` box in dashboard granules page?  The search field is actually `infix`, matches part of the granuleId.

jennyhliu replied:
nvm.  I see `contains` is used for the db search.

jennyhliu commented on packages/api/models/granules.js:
```diff
@@ -446,16 +420,72 @@ class Granule extends Manager {
+      filterArray,
+      filterExpression: (filterArray.length > 0) ? filterArray.join(' AND ') : undefined,
+      keyConditionArray,
+      keyExpression: keyConditionArray.join(' AND '),
```
`keyExpression` is not used

jennyhliu commented on packages/api/models/granules.js:
```diff
@@ -446,16 +420,72 @@ class Granule extends Manager {
+      attributeNames,
+      attributeValues,
+      filterExpression,
+      keyConditionArray,
```
use `keyExpression` instead?

dopeters replied:
Stopped `keyExpression` from being returned in `getDynamoDbSearchParams`. Want to use `keyConditionArray` here because we end up adding the collectionId to that array later in this function

`keyConditionArray.push('#collectionId = :collectionId');`

jennyhliu reviewed (changes requested):
Looks good!
We can also add search parameters to https://github.com/nasa/cumulus/blob/master/example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js#L608
and verify that the selected granules/collection are reported.

jennyhliu commented:
https://github.com/nasa/cumulus/blob/master/packages/api/endpoints/reconciliation-reports.js#L122
You probably need to update the function which validates the input parameters.
Api doc needs to be updated.  https://github.com/nasa/cumulus-api

jennyhliu reviewed (approved):
Nice job! LGTM

jennyhliu reviewed (changes requested):
If I change the granuleId parameter to granuleId: 'MYD13Q1', in the integration test, the report doesn't have anything

jennyhliu commented on packages/api/lib/reconciliationReport.js:
```diff
@@ -117,6 +117,36 @@ function convertToDBGranuleSearchParams(params) {
+function convertToDBScanGranuleSearchParams(params) {
+  const {
+    collectionId,
+    granuleId: granuleIdParam,
```
There won't be `granuleId` parameter after normalizeEvent, only `granuleIds`

dopeters replied:
Ahhh yeah I see that now. Thanks, will adjust

jennyhliu replied:
Can you add a test case for the lambda `packages/api/tests/lambdas/test-create-reconciliation-report.js`?

dopeters replied:
Absolutely!

dopeters commented:
> I think it should be `granuleId`.
> Can you also add `status` here and lambda test?
> Don't forget api doc.

Yep, you're right. Fixed.
Added status as a param.
Opened a PR for the docs here: https://github.com/nasa/cumulus-api/pull/295

dopeters commented:
Dashboard side, we were passing `granuleIds` instead of `granuleId` as the param, so opened up a PR to fix that https://github.com/nasa/cumulus-dashboard/pull/866
